mqtt/mqtt-edgehub/src/auth/authorization/edgehub.rs

```diff
@@ -119,6 +122,64 @@ impl EdgeHubAuthorizer {
             .iter()
             .any(|allowed_topic| topic.starts_with(allowed_topic))
     }
+
+    fn get_on_behalf_of_id(topic: &str) -> Option<String> {
+        // topics without the new topic format cannot have on_behalf_of_ids
+        if !topic.starts_with("$iothub/clients") {
+            return None;
+        }
+        let topic_parts = topic.split('/').collect::<Vec<&str>>();
+        let device_id = topic_parts.get(2);
+        let module_id = match topic_parts.get(3) {
+            Some(s) => {
+                if *s == "modules" {
+                    topic_parts.get(4)
+                } else {
+                    None
+                }
+            }
+            None => None,
+        };
+        match module_id {
+            Some(mid) => match device_id {
+                Some(did) => Some(format!("{}/{}", did, mid)),
+                None => None,
+            },
+            None => match device_id {
+                Some(id) => Some((*id).to_string()),
+                None => None,
+            },
+        }
+    }
+
+    fn check_authorized_cache(&self, client_id: &ClientId, topic: &str) -> bool {
+        let on_behalf_of_id = if let Some(id) = EdgeHubAuthorizer::get_on_behalf_of_id(topic) {
```

Comment:
the code here is a bit difficult to read. I guess it is because of too many nesting levels. try to flatten it somehow

Reply:
Restructured this part based on Vadim's suggestion